`resolve.tsconfig` with `true` or relative path does not find `tsconfig.json`

### Have you used AI?

No

### Bug Description

Trying to use [`resolve.tsconfig`](https://webpack.js.org/configuration/resolve/#resolvetsconfig) to set up some aliases.

However, using `resolve.tsconfig: true` is leading to an error:

```bash
$ nix# npm --prefix assets run deploy

> deploy
> webpack --mode production

assets by status 639 bytes [cached] 1 asset
./js/App.tsx 230 bytes [built] [code generated]

ERROR in ./js/App.tsx 1:0-26
Module not found: Error: ENOENT: no such file or directory, open '/home/me/dev/playground/elixir/phoenix/myapp/assets/js/tsconfig.json'

ERROR in ./js/App.tsx 2:0-46
Module not found: Error: ENOENT: no such file or directory, open '/home/me/dev/playground/elixir/phoenix/myapp/assets/js/tsconfig.json'

webpack 5.106.2 compiled with 2 errors in 782 ms
```

Same thing with relative path, only `tsconfig: path.resolve(import.meta.dirname, "tsconfig.json")` works.

### Link to Minimal Reproduction and step to  reproduce

https://github.com/datafoo/webpack-issue-resolve.tsconfig

Reproduction steps

- Pull up developement environment with nix

  ```bash
  nix develop
  ```

- Install npm dependencies

  ```bash
  npm --prefix assets install
  ```

- Notice the deploy failure due to `resolve.tsconfig: true`

  ```bash
  npm --prefix assets run deploy
  ```

### Expected Behavior

It should find `tsconfig.json` just like it does when using `tsconfig: path.resolve(import.meta.dirname, "tsconfig.json"),`:

```bash
$ nix# npm --prefix assets run deploy

> deploy
> webpack --mode production

asset app.js 185 KiB [emitted] [minimized] (name: app) 1 related asset
modules by path ./node_modules/react-dom/ 533 KiB
  modules by path ./node_modules/react-dom/*.js 2.67 KiB
    ./node_modules/react-dom/client.js 1.34 KiB [built] [code generated]
    ./node_modules/react-dom/index.js 1.33 KiB [built] [code generated]
  modules by path ./node_modules/react-dom/cjs/*.js 530 KiB
    ./node_modules/react-dom/cjs/react-dom-client.production.js 523 KiB [built] [code generated]
    ./node_modules/react-dom/cjs/react-dom.production.js 6.5 KiB [built] [code generated]
modules by path ./node_modules/react/ 17 KiB
  ./node_modules/react/index.js 186 bytes [built] [code generated]
  ./node_modules/react/cjs/react.production.js 16.8 KiB [built] [code generated]
modules by path ./node_modules/scheduler/ 10.1 KiB
  ./node_modules/scheduler/index.js 194 bytes [built] [code generated]
  ./node_modules/scheduler/cjs/scheduler.production.js 9.94 KiB [built] [code generated]
./js/App.tsx 230 bytes [built] [code generated]
webpack 5.106.2 compiled successfully in 3600 ms
```

### Actual Behavior

```bash
$ nix develop

$ nix# npm --prefix assets install

added 424 packages, and audited 427 packages in 12s

55 packages are looking for funding
  run `npm fund` for details

2 high severity vulnerabilities

To address all issues, run:
  npm audit fix

Run `npm audit` for details.

$ nix# npm --prefix assets run deploy

> deploy
> webpack --mode production

assets by status 639 bytes [cached] 1 asset
./js/App.tsx 230 bytes [built] [code generated]

ERROR in ./js/App.tsx 1:0-26
Module not found: Error: ENOENT: no such file or directory, open '/home/me/dev/playground/elixir/phoenix/myapp/assets/js/tsconfig.json'

ERROR in ./js/App.tsx 2:0-46
Module not found: Error: ENOENT: no such file or directory, open '/home/me/dev/playground/elixir/phoenix/myapp/assets/js/tsconfig.json'

webpack 5.106.2 compiled with 2 errors in 782 ms

$ nix# node --version
v24.15.0

$ nix# npm --version
11.12.1

```

### Environment

```shell
System:
  OS: Linux 6.12 cpe:/o:nixos:nixos:25.11 25.11 (Xantusia)
  CPU: (6) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  Memory: 18.87 GB / 31.29 GB
Binaries:
  Node: 24.15.0 - /nix/store/chsxa1iw7s3m51v4kj43yd1nw26xvi21-nodejs-24.15.0/bin/node
  npm: 11.12.1 - /nix/store/chsxa1iw7s3m51v4kj43yd1nw26xvi21-nodejs-24.15.0/bin/npm
Browsers:
  Firefox: 150.0.1
  Firefox Developer Edition: 150.0.1
Packages:
  babel-loader: ^10.1.1 => 10.1.1 
  css-loader: ^7.1.4 => 7.1.4 
  css-minimizer-webpack-plugin: ^7.0.4 => 7.0.4 
  sass-loader: ^16.0.7 => 16.0.7 
  terser-webpack-plugin: ^5.5.0 => 5.5.0 
  webpack: ^5.106.2 => 5.106.2 
  webpack-cli: ^7.0.2 => 7.0.2
```

### Is this a regression?

None

### Last Working Version

_No response_

### Additional Context

_No response_

--- a/.changeset/tsconfig-true-upward-traversal.md
+++ b/.changeset/tsconfig-true-upward-traversal.md
@@ -1,0 +1,5 @@
+---
+"enhanced-resolve": patch
+---
+
+When `tsconfig: true` is used, walk up parent directories to find `tsconfig.json`, matching TypeScript's own `findConfigFile` behavior.

--- a/lib/TsconfigPathsPlugin.js
+++ b/lib/TsconfigPathsPlugin.js
@@ -337,6 +337,32 @@ module.exports = class TsconfigPathsPlugin {
 			return callback(null, cached);
 		}
 
+		if (this.isAutoConfigFile) {
+			this._findTsconfigUpward(
+				resolver,
+				request.path || process.cwd(),
+				(err, result) => {
+					if (err) {
+						request.tsconfigPathsMap = null;
+						return callback(err);
+					}
+					if (!result) {
+						request.tsconfigPathsMap = null;
+						return callback(null, null);
+					}
+					const map = /** @type {TsconfigPathsMap} */ (result);
+					request.tsconfigPathsMap = map;
+					if (resolveContext.fileDependencies) {
+						for (const fileDependency of map.fileDependencies) {
+							resolveContext.fileDependencies.add(fileDependency);
+						}
+					}
+					callback(null, map);
+				},
+			);
+			return;
+		}
+
 		const absTsconfigPath = resolver.join(
 			request.path || process.cwd(),
 			this.configFile,
@@ -344,12 +370,6 @@ module.exports = class TsconfigPathsPlugin {
 		this._loadTsconfigPathsMap(resolver, absTsconfigPath, (err, result) => {
 			if (err) {
 				request.tsconfigPathsMap = null;
-				if (
-					this.isAutoConfigFile &&
-					/** @type {NodeJS.ErrnoException} */ (err).code === "ENOENT"
-				) {
-					return callback(null, null);
-				}
 				return callback(err);
 			}
 
@@ -362,6 +382,45 @@ module.exports = class TsconfigPathsPlugin {
 			}
 			callback(null, map);
 		});
+	}
+
+	/**
+	 * Walk up from startDir to the filesystem root looking for tsconfig.json.
+	 * Like TypeScript's own `findConfigFile` / `forEachAncestorDirectory`.
+	 * @param {Resolver} resolver the resolver
+	 * @param {string} startDir the directory to start searching from
+	 * @param {(err: Error | null, result?: TsconfigPathsMap | null) => void} callback the callback
+	 * @returns {void}
+	 */
+	_findTsconfigUpward(resolver, startDir, callback) {
+		const { fileSystem } = resolver;
+		const configFileName = this.configFile;
+
+		/**
+		 * @param {string} dir current directory
+		 */
+		const check = (dir) => {
+			const candidate = resolver.join(dir, configFileName);
+			fileSystem.stat(candidate, (statErr) => {
+				if (!statErr) {
+					// Found — load it
+					this._loadTsconfigPathsMap(resolver, candidate, (loadErr, result) => {
+						if (loadErr) return callback(loadErr);
+						callback(null, result);
+					});
+					return;
+				}
+				// Not found — move to parent
+				const parentDir = resolver.dirname(dir);
+				if (parentDir === dir) {
+					// Reached filesystem root, no tsconfig.json found
+					return callback(null, null);
+				}
+				check(parentDir);
+			});
+		};
+
+		check(startDir);
 	}
 
 	/**

--- a/test/fixtures/tsconfig-paths/upward-traversal/src/app/index.ts
+++ b/test/fixtures/tsconfig-paths/upward-traversal/src/app/index.ts
@@ -1,0 +1,1 @@
+import { helper } from "@utils/helper";

--- a/test/fixtures/tsconfig-paths/upward-traversal/src/utils/helper.ts
+++ b/test/fixtures/tsconfig-paths/upward-traversal/src/utils/helper.ts
@@ -1,0 +1,1 @@
+export const helper = "helper";

--- a/test/fixtures/tsconfig-paths/upward-traversal/tsconfig.json
+++ b/test/fixtures/tsconfig-paths/upward-traversal/tsconfig.json
@@ -1,0 +1,8 @@
+{
+  "compilerOptions": {
+    "baseUrl": ".",
+    "paths": {
+      "@utils/*": ["./src/utils/*"]
+    }
+  }
+}

--- a/test/tsconfig-paths.test.js
+++ b/test/tsconfig-paths.test.js
@@ -1268,6 +1268,65 @@ describe("TsconfigPathsPlugin", () => {
 		});
 	});
 
+	describe("tsconfig: true should walk up parent directories", () => {
+		const upwardDir = path.resolve(
+			__dirname,
+			"fixtures",
+			"tsconfig-paths",
+			"upward-traversal",
+		);
+
+		it("should find tsconfig.json in parent directory when resolving from subdirectory", (done) => {
+			const resolver = ResolverFactory.createResolver({
+				fileSystem,
+				extensions: [".ts", ".tsx"],
+				mainFields: ["browser", "main"],
+				mainFiles: ["index"],
+				tsconfig: true,
+			});
+
+			// resolve from src/app/ — tsconfig.json is in upward-traversal/ (parent's parent)
+			resolver.resolve(
+				{},
+				path.join(upwardDir, "src", "app"),
+				"@utils/helper",
+				{},
+				(err, result) => {
+					if (err) return done(err);
+					if (!result) return done(new Error("No result"));
+					expect(result).toEqual(
+						path.join(upwardDir, "src", "utils", "helper.ts"),
+					);
+					done();
+				},
+			);
+		});
+
+		it("should still fall through when no tsconfig.json exists anywhere up", (done) => {
+			const noTsconfigDir = path.resolve(
+				__dirname,
+				"fixtures",
+				"tsconfig-paths",
+				"no-tsconfig",
+			);
+
+			const resolver = ResolverFactory.createResolver({
+				fileSystem,
+				extensions: [".ts", ".tsx"],
+				mainFields: ["browser", "main"],
+				mainFiles: ["index"],
+				tsconfig: true,
+			});
+
+			resolver.resolve({}, noTsconfigDir, "./src/index", {}, (err, result) => {
+				if (err) return done(err);
+				if (!result) return done(new Error("No result"));
+				expect(result).toEqual(path.join(noTsconfigDir, "src", "index.ts"));
+				done();
+			});
+		});
+	});
+
 	describe("JSONC support (comments in tsconfig.json)", () => {
 		const jsoncExampleDir = path.resolve(
 			__dirname,

